### ISSUE TITLE
The Iterator of ServerListManager has some concurrency problem in 1.x client 

### ISSUE THREAD

**Issue by gaofeihang** [alibaba/nacos#8203]
**Describe the bug**
The operation of updating currentServerAddr and getting nextServerAddr is not thread safe. It may cause `NoSuchElementException` when `Iterator.next()` is called. 

One situation is as follows:

Nacos config client starts `LongPollingRunnable` task and will call `ServerHttpAgent` in async thread. If the business logic calls `NacosConfigService.getConfig()` and `ServerHttpAgent` concurrently, a problem about the Iterator of `ServerListManager` may occur.

ServerHttpAgent calls `ServerListManager.getCurrentServerAddr()` every time when a http request is sent. The currentServerAddr field is initialized by an iterator.

```
private volatile String currentServerAddr;
private Iterator<String> iterator;

...

public String getCurrentServerAddr() {
    if (StringUtils.isBlank(currentServerAddr)) {
        iterator = iterator();
        currentServerAddr = iterator.next();
    }
    return currentServerAddr;
}
```

When two threads call getCurrentServerAddr() and enter the judgement `StringUtils.isBlank(currentServerAddr)` concurrently, let's suppose the next steps happen as follows

1. Thread t1 initialized the iterator field with value v1
2. Thread t2 initialized the iterator field with value v2
3. Thread t2 calls v2.next()
4. Thread t1 recognized the new value of the iterator field `v2` and calls v2.next()

Iterator v2 has been called twice. If the `serverUrls`  and the iterator have only one sever address, when nginx and domain address is used for example, the second invocation of v2.next() will throw `NoSuchElementException`

The other operations with the iterator also have risks. Such as the getNextServer operation in `ServerHttpAgent`

```
if (serverListMgr.getIterator().hasNext()) {
    currentServerAddr = serverListMgr.getIterator().next();
}
```

If two threads enter the judgement at the same time, `Iterator.next()` may cause `NoSuchElementException`.

**Expected behavior**
`ServerListManager.getCurrentServerAddr()` can work without Exceptions when it is called concurrently.

**Acutally behavior**
`ServerListManager.getCurrentServerAddr()` throws `NoSuchElementException` when it is called concurrently.

**How to Reproduce**
Steps to reproduce the behavior: Run the test codes below

```
public class ServerListMangerTest {

    private static final Logger LOG = LoggerFactory.getLogger(ServerListMangerTest.class);

    @Test
    public void testCurrentServerAddr() {

        int concurrency = 10;
        int maxAttempt = 3;

        AtomicBoolean failed = new AtomicBoolean(false);

        ExecutorService executorService = new ThreadPoolExecutor(
                concurrency, concurrency, 60, TimeUnit.SECONDS, new LinkedBlockingQueue<>());

        for (int i = 0; i < maxAttempt; i++) {
            final int attempt = i;

            // create a new ServerListManager instance to make the currentServerAddr field not initialized
            final ServerListManager serverListManager = new ServerListManager(Arrays.asList("1.2.3.4:8848"));

            final CountDownLatch latch = new CountDownLatch(concurrency);

            // imitate the situation that ServerHttpAgent.httpGet() is called concurrently
            for (int j = 0; j < concurrency; j++) {
                executorService.execute(new Runnable() {
                    @Override
                    public void run() {
                        if (failed.get()) {
                            return;
                        }

                        try {
                            serverListManager.getCurrentServerAddr();
                        } catch (NoSuchElementException e) {
                            failed.set(true);
                            LOG.error("failed at " + attempt + " attempt", e);
                        } finally {
                            latch.countDown();
                        }
                    }
                });
            }

            try {
                latch.await(1000, TimeUnit.MILLISECONDS);
            } catch (Exception e) {
                LOG.error("await error", e);
            }

            if (failed.get()) {
                break;
            }
        }

        executorService.shutdown();

        Assert.assertFalse(failed.get());
    }

}
```

**Desktop (please complete the following information):**
 - OS: Mac OSX
 - Version: nacos-client 1.4.3
 - Module: config
 - SDK: original

**Additional context**
None.

**Comment by KomachiSion** [comment]
I remember this problem has been fixed in 2.X. Can you have a research and fix it with same way as 2.X?

**Comment by gaofeihang** [comment]
> I remember this problem has been fixed in 2.X. Can you have a research and fix it with same way as 2.X?

I took a short research yesterday and found `ServerListManager.getCurrentServerAddr()` method has no change in 2.x. See https://github.com/alibaba/nacos/blob/2.1.0-BETA/client/src/main/java/com/alibaba/nacos/client/config/impl/ServerListManager.java#L465

I thought it was because 2.x uses gRPC as the main transport approach and `ServerHttpAgent` is no longer frequently used. I thought this problem only needed to be fixed in 1.x.

After your reminder, I took a research again. Here are some new conclusions.

1. `ServerListManager.getCurrentServerAddr()` is not used by gRPC transport, a new method `ServerListManager.getNextServerAddr()` is used instead.

```
public String getNextServerAddr() {
    if (iterator == null || !iterator.hasNext()) {
        refreshCurrentServerAddr();
        return currentServerAddr;
    }
    try {
        return iterator.next();
    } catch (Exception e) {
        //No nothing.
    }
    refreshCurrentServerAddr();
    return currentServerAddr;    
}
```
2. The new method fix the iterator problem after the initialization of `iterator` by catching `NoSuchElementException` and resetting the iterator. But the problem still exists in `refreshCurrentServerAddr()`, when `iterator` is not initialized for example. 
3. After changing the test method from `getCurrentServerAddr()` to `getNextServerAddr()` in the test codes, the problem can also be reproduced. It should be noted that the problem can be produced in a small probability, not every time.
4. Now I think 2.x also has the same problem. `refreshCurrentServerAddr()` needs to  be synchronized if  a fix plan with least  changes is wanted.

I will optimize the PR by migrating the `getNextServerAddr()` method to 1.x and fix the problem in 1.x first.
